raco/backends/myria/myria.py

```diff
@@ -1721,6 +1824,193 @@ def fire(self, op):
         return algebra.UnionAll(children)
 
 
+class FillInJoinPullOrder(rules.Rule):
+
+    def __init__(self, pull_order):
+        self._disabled = False
+        self.pull_order = pull_order
+
+    def fire(self, op):
+        if not isinstance(op, MyriaSymmetricHashJoin):
+            return op
+        op.pull_order = self.pull_order
+        return op
+
+
+class PropagateAsyncFTBuffer(rules.Rule):
+
+    def __init__(self):
+        self._disabled = False
+
+    def fire(self, op):
+        if not isinstance(op, MyriaIDBController):
+            return op
+        buffer_type = op.get_agg_json()
+        if buffer_type['type'] == "CountFilter":
+            # do not propagate CountFilter
+            return op
+        # only do it for the iterative input
+        if isinstance(op.args[1], MyriaShuffleConsumer):
+            op.args[1].input.buffer_type = buffer_type
+        return op
+
+
+class ReplaceStoreIDB(rules.Rule):
+
+    @staticmethod
+    def collect_and_replace(op, idbproducers):
+        if not isinstance(op, algebra.NaryOperator):
+            return
+        if isinstance(op, algebra.UntilConvergence):
+            for idbproducer in op.children():
+                if isinstance(idbproducer.input, MyriaIDBController):
+                    idbproducers[idbproducer.input.name] = idbproducer
+            return
+        newchildren = []
+        for child in op.children():
+            if isinstance(child, algebra.UntilConvergence):
+                ReplaceStoreIDB.collect_and_replace(child, idbproducers)
+            if isinstance(child, algebra.Store) and \
+                    isinstance(child.input, algebra.ScanIDB):
+                assert child.input.name in idbproducers
+                idbproducers[child.input.name].input.relation_key = \
+                    child.relation_key
+            else:
+                newchildren.append(child)
```

Comment:
It could be any operator except `Store`.  We start from the whole MyriaL program and check the children of the current operator. If the child is a `Store` on top of a `ScanIDB`, we discard this store and tell the IDBController to store directly. Otherwise the child, no matter which type it is, is kept.